Add docs for multi-cluster Kiali setup (#157)
openshift-service-mesh/sail-operator: docs/ossm/kiali-multi-cluster/README.md

[Return to OSSM Docs](../)

# Kiali - multi-cluster

For multi-cluster Istio deployments, Kiali can show you a single unified view of your mesh across clusters.

Before proceeding with the setup, ensure you meet the requirements.

## Requirements

> - Two openshift clusters. In this tutorial they are named `east` and `west`.

Reviewer: correct capitalization in OpenShift
```suggestion
- Two OpenShift clusters. In this tutorial they are named `east` and `west`.
```

- Cluster Admin for each cluster.
- Kiali Operator v1.89 installed on the `east` cluster.

## Setup

In this tutorial, we will deploy Kiali on the `east` cluster and then grant Kiali access to the `west` cluster. The unified multi-cluster setup requires the Kiali Service Account (SA) to have read access to each Kubernetes cluster in the mesh. This is separate from the user credentials that are required when a user logs into Kiali. Kiali uses the user's credentials to check if the user has access to a namespace and when performing any write operation such as creating/editing/deleting objects in Kubernetes. To give the Kiali Service Account access to each remote cluster, a kubeconfig with credentials needs to be created and mounted into the Kiali pod.

### Procedure

1. Install Kiali on the `east` cluster.

   Create a file named `kiali.yaml`.

   ```yaml
   apiVersion: kiali.io/v1alpha1
   kind: Kiali
   metadata:
     name: kiali
     namespace: istio-system
   spec:
     version: v1.89
   ```

   Apply the yaml file into the `east` cluster.

   ```sh
   oc --context east apply -f kiali.yaml
   ```

   Wait for Kiali to become ready.

   ```sh

>    oc --context east wait --for=condition=Successful --timeout=60s kialis/kiali -n istio-system

Reviewer: I actually would wait longer than 60s. On slow clusters, I've seen it take over a minute. Perhaps say 2m here:
```suggestion
   oc --context east wait --for=condition=Successful --timeout=2m kialis/kiali -n istio-system
```

   ```

1. Create an `OAuthClient` on the remote cluster so that Kiali can access the OpenShift API server on behalf of users.

>    Find your Kiali URL

Reviewer: This isn't really finding the URL, is it? "spec.host" I think is the route hostname (i.e. it doesn't include the protocol "http" or port number here).
```suggestion
   Find your Kiali route's hostname
```

   ```sh
   oc --context east get route kiali -n istio-system -o jsonpath='{.spec.host}'
   ```

   Create a file named `oauthclientwest.yaml`

   ```yaml
   apiVersion: oauth.openshift.io/v1
   grantMethod: auto
   kind: OAuthClient
   metadata:
     labels:
       app: kiali
       app.kubernetes.io/instance: kiali
       app.kubernetes.io/name: kiali
       app.kubernetes.io/part-of: kiali
     name: kiali-istio-system
   redirectURIs:

>      - https://<your-kiali-route>/api/auth/callback/west

Reviewer: This isn't the route per se, its the hostname retrieved in the command from above:
```suggestion
     - https://<your-kiali-route-hostname>/api/auth/callback/west
```

   ```

   Create the `OAuthClient` in the west cluster.

   ```sh
   oc --context west apply -f oauthclientwest.yaml
   ```

1. Create a remote cluster secret.

   In order to access a remote cluster, you must provide a kubeconfig to Kiali via a Kubernetes secret. You can use [this script](https://raw.githubusercontent.com/kiali/kiali/master/hack/istio/multicluster/kiali-prepare-remote-cluster.sh) to simplify this process for you. Running this script will:

   - Create a Service Account for Kiali in the remote cluster.
   - Create RBAC resources for this Service Account in the remote cluster.
   - Create a kubeconfig file and save this as a secret in the namespace where Kiali is deployed on the `east` cluster.

>    1. Download the `kiali-prepare-remote-cluster` script.

Reviewer: ```suggestion
   1. Download the `kiali-prepare-remote-cluster.sh` script.
```

      ```sh
      curl -L -o kiali-prepare-remote-cluster.sh https://raw.githubusercontent.com/kiali/kiali/master/hack/istio/multicluster/kiali-prepare-remote-cluster.sh
      ```

   2. Make the script executeable.

      ```sh
      chmod +x kiali-prepare-remote-cluster.sh
      ```

   3. Run the script passing your `east` and `west` cluster contexts.

      ```sh

>       ./kiali-prepare-remote-cluster.sh --kiali-cluster-context east --remote-cluster-context west --view-only false --kiali-resource-name kiali --remote-cluster-namespace istio-system --remote-cluster-name west

Reviewer: In my 2-cluster setup, I got this error message after running this (which makes sense):
```
ERROR: The remote cluster named [api-mazzwest-lab-upshift-rdu2-redhat-com:6443] in the remote context is configured with insecure-skip-tls-verify. If you agree to use that, pass in '--allow-skip-tls-verify true'. Aborting.
```

So I had to re-run the script and pass in the extra arg `--allow-skip-tls-verify true` 

Not sure if you want to mention this or not here in the docs. It was pretty self-explanatory when it happened, so at least the script's error message was helpful.

Author: Eh I probably won't mention it because I expect people's real clusters to have trusted certs already and like you mentioned the error is pretty self explanatory if they don't.

      ```

   **Note:** Use the option `--help` for additional details on how to use the script.

1. Restart Kiali to pickup the remote secret.

   ```sh
   oc --context east rollout restart deployments/kiali -n istio-system
   ```

>    Wait for Kiali to become ready

Reviewer: Missing punctuation - just to be consistent with everything else that has periods at the end
```suggestion
   Wait for Kiali to become ready.
```

   ```sh
   oc --context east rollout status deployments/kiali -n istio-system
   ```

1. Login to Kiali.

   When you first visit Kiali, you will login to the cluster where Kiali is deployed. In our case it will be the `east` cluster.

   ```sh

>    oc --context east get route -l app=kiali -n istio-system

Reviewer: Kinda confused what this command was for. I think you mean for this to show where the user should point his browser. Perhaps something more specific - this opens a browser tab right to Kiali, maybe use this:

```
xdg-open "https://$(oc --context east get route -l app=kiali -n istio-system -o jsonpath='{..spec.host}')"
```

Reviewer: Oh, also, I try not to use "app" label in my official docs because technically we should be telling people to use the `app.kubernetes.io/` labels ("app" is unofficially deprecated - we want to get people to use the standard names). So..

```
xdg-open "https://$(oc --context east get route -l app.kubernets.io/name=kiali -n istio-system -o jsonpath='{..spec.host}')"
```

Author: I may just say "open your browser to..." here to be more platform agnostic.

   ```

1. Login to the `west` cluster through Kiali.

>    In order to see other clusters in the Kiali UI, you must first login as a user to those clusters through Kiali. Click on the user profile dropdown in the top right hand menu. Then select `Login to west`. You will again be redirected to an openshift login page and prompted for credentials but this will be for the `west` cluster.

Reviewer: OpenShift capitalization

```suggestion
   In order to see other clusters in the Kiali UI, you must first login as a user to those clusters through Kiali. Click on the user profile dropdown in the top right hand menu. Then select `Login to west`. You will again be redirected to an OpenShift login page and prompted for credentials but this will be for the `west` cluster.
```

1. Verify that Kiali shows information from both clusters.

>    1. Navigate to the `Overview` page from the lefthand nav and verify you can see namespaces from both clusters.
> 
>    1. Navigate to the `Mesh` page from the lefthand nav and verify you see both clusters on the mesh graph.

Reviewer: On line 131 above, you have "right hand menu". Here you have "lefthand nav"...
`right hand` (with a space) and `lefthand` (without a space). I don't know the way the docs people want this, but we should be consistent, one way or the other. I'll suggest using space here:
```suggestion
   1. Navigate to the `Overview` page from the left hand nav and verify you can see namespaces from both clusters.

   1. Navigate to the `Mesh` page from the left hand nav and verify you see both clusters on the mesh graph.
```

- Cluster Admin for each cluster.
- Kiali Operator v1.89 installed on the `east` cluster.

## Setup

In this tutorial, we will deploy Kiali on the `east` cluster and then grant Kiali access to the `west` cluster. The unified multi-cluster setup requires a Kiali Service Account (SA) to have read access to each Kubernetes cluster in the mesh. This is separate from the user credentials that are required when a user logs into Kiali. Kiali uses the user's credentials to check if the user has access to a namespace and when performing any write operation such as creating/editing/deleting objects in Kubernetes. To give the Kiali Service Account access to each remote cluster, a kubeconfig with credentials needs to be created and mounted into the Kiali pod.

### Procedure

1. Install Kiali on the `east` cluster.

   Create a file named `kiali.yaml`.

   ```yaml
   apiVersion: kiali.io/v1alpha1
   kind: Kiali
   metadata:
     name: kiali
     namespace: istio-system
   spec:

>      version: v1.89

Reviewer 2: This shouldn't be version: default?

   ```

   Apply the yaml file into the `east` cluster.

   ```sh
   oc --context east apply -f kiali.yaml
   ```

   Wait for Kiali to become ready.

   ```sh
   oc --context east wait --for=condition=Successful --timeout=2m kialis/kiali -n istio-system
   ```

1. Create an `OAuthClient` on the remote cluster so that Kiali can access the OpenShift API server on behalf of users.

   Find your Kiali route's hostname.

   ```sh
   oc --context east get route kiali -n istio-system -o jsonpath='{.spec.host}'
   ```

   Create a file named `oauthclientwest.yaml`

   ```yaml
   apiVersion: oauth.openshift.io/v1
   grantMethod: auto
   kind: OAuthClient
   metadata:
     labels:
       app: kiali
       app.kubernetes.io/instance: kiali
       app.kubernetes.io/name: kiali
       app.kubernetes.io/part-of: kiali
     name: kiali-istio-system
   redirectURIs:

>      - https://<your-kiali-route-hostname>/api/auth/callback/west

Author: This is the `OAuthClient` that gets created in the `west` cluster that redirects your west cluster login back to Kiali in the `east` cluster. It needs to have `west` appended on the end of the url since that tells the Kiali server which cluster the redirect is for.
